### ISSUE TITLE
pip_requirements: Renovate does not parse unbound requirements

### ISSUE THREAD

**Issue by alecgerona** [renovatebot/renovate#6955]
I've toggled my project on the dashboard with a repository that has this settings file already: `renovate.json`
```json
{
  "extends": [
    "config:base",
    ":githubComToken(XXXX)"
  ],
  "pip_requirements": {
    "fileMatch": [
      "^requirements/.*\\.txt$"
    ]
  }
}
```
The webhook is created, user is added with the proper permissions etc. My requirement files are all under the `requirements` folder (base.txt, etc.). I've let it sit for a few days since I know about the existing issues Renovate has with Gitlab. But there's still no MR to onboard or pin my requirements (I've left them blank to let Renovate deal with them) even though jobs are being executed on time (I've also seen some updates on my gitlab-ci file and dockerfiles but still no MRs. The most recent job id to reflect my latest settings change is #214805857. 

I've also confirmed that this isn't a stale repository since the initial commit was just pushed last week on a private organization in Gitlab. Hope this is just a silly mistake on my end.

**Comment by rarkins** [comment]
I have renamed and transferred to the main repo. To resolve this issue, we should parse/detect the requirements, although still not pin them. At least then the files and dependencies will all show up in logs.

Currently we do not pin *unbound* requirements in any language (including ones that allow for a value like `*`), so that needs to be a separate feature to consider, keeping in mind backwards compatibility. We only pin them if there is some range already existing, e.g. `1.*`

**Comment by alecgerona** [comment]
Awesome. Let me know if I can contribute anything for this.

**Comment by rarkins** [comment]
Sure, you're welcome to submit a PR. We'd need a tweaking of our regex, it's defined here: https://github.com/renovatebot/renovate/blob/6ddda0fa4b95b3090d08fba9ce26db4f9cf640b1/lib/manager/pip_requirements/extract.ts#L15

**Comment by alecgerona** [comment]
Will definitely check this out. Can I be invited to the Slack channel for more info?

**Comment by rarkins** [comment]
Sure, send an email so we can invite you

**Comment by alecgerona** [comment]
Cool. I sent you an email.

**Comment by alecgerona** [comment]
Might be related to this, Renovate also doesn't parse requirements that has a version that doesn't follow semantic versioning. Great example is [black](https://github.com/psf/black) code formatter. Its version is XX.YbZ.

**Comment by rarkins** [comment]
@alecgerona you may be able to get it to work by using a package rule to configure `loose` versioning for that package.

**Comment by alecgerona** [comment]
With something like
```json
"packageRules": [
    {
      "packageNames": ["black"],
      "versioning": "loose"
    }
  ]
```
? 
Will give it a try.

**Comment by alecgerona** [comment]
Hi @rarkins, been getting this:
```
{
            "depName": "black",
            "updates": [],
            "depIndex": 0,
            "warnings": [],
            "datasource": "pypi",
            "skipReason": "unsupported-value",
            "fromVersion": "19.10b0",
            "currentValue": "==19.10b0"
},
```

**Comment by rarkins** [comment]
Sorry, that makes sense. Loose versioning doesn't understand the == part as it's python-specific. I'm not sure why we reject this version string with our default pep440 versioning. Anyway, it's unrelated to this issue so I'd prefer not to complicate it with further discussion. Please raise a new issue for us to look into supporting this syntax with out pep440 versioning scheme.

**Comment by jab** [comment]
Looks like this issue affects the `pip-compile` manager as well, not just `pip-requirements` (hi from https://github.com/renovatebot/renovate/discussions/12432 :). Does that perhaps bump priority on getting a fix rolled out?

**Comment by rarkins** [comment]
PRs welcome

**Comment by AIGeneratedUsername** [comment]
What is a current workaround?

Imagine, I have a file with unpinned packages:

```
# requirements.txt

pandas
numpy
redis
```

Do I need to pin all my packages manually to make Renovatebot working until the issue is not solved?

**Comment by PhilipAbed** [comment]
python works in an environment where dependencies are installed, 
Renovate works with files in a repository,

The `requirements.txt` file is where you save your environment

if your file doesn't have a version, that means every time you run `pip install -r requirements.txt`
the last version will be automatically installed 
then you dont need renovate to do that for you

**Comment by AIGeneratedUsername** [comment]
> 

No, you I think you did not understand correctly what was asked here. People want Renovate to overwrite the "requirements.txt" file with pinned versions.

**Comment by rarkins** [comment]
I am not sure renovate pins versions for any manager when no constraint is set. Only when there's an existing one, we have traditionally left inbound dependencies as is

**Comment by PhilipAbed** [comment]
@AIGeneratedUsername 
in order for renovate to PIN the dependency
it needs to have a range like

```
foobar == '^1.0.0'
spameggs == '<1.3.0'
```

but if  you want a renovate `type:replace` to create an Update, 
that update should have `CurrentVersion` to replace -> `NewVersion` o replace it with
in this case there's no CurrentVersion, so obviously something will go wrong.


Thing is i still don't get why would you need such functionality when `requirements.txt` itself provides that for you when u don't set a version.

i'm not an expert on regex managers but 
how about using `regexManagers` to do it? 
looking for file match: `requirements.txt`
matchStrings: do a regex for a dependency name no spaces/additions 
and use the pypi datasource template

**Comment by rarkins** [comment]
@PhilipAbed I am not sure that range syntax is correct for pep440 but the concept is ok

**Comment by PhilipAbed** [comment]
i think we could close this issue as there's nothing to fix from our side, maybe move it to discussions?
